Add support for Ace of Spades TKL keyboard

rys opened the pull request:
Adds support for the Ace of Spades TKL custom keyboard PCB. It's a custom limited-run TKL with fixed layout, designed by tpk and manufactured by Salvun.

## Description

Default keymap for the fixed ISO layout, plus VIA support and a personal keymap for aholland909 (the PCB's designer) in separate keymaps. All 3 maps have been tested on a production PCB, and VIA support was tested with VIA v1.2.7.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

N/A.

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

zvecr commented on keyboards/tpk/ace_of_spades/ace_of_spades.h:
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include "quantum.h"
+#include "config_common.h"
```
```suggestion
```

noroadsleft commented on keyboards/tpk/ace_of_spades/rules.mk:
```diff
@@ -0,0 +1,35 @@
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
```
```suggestion
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
```

noroadsleft commented on keyboards/tpk/ace_of_spades/rules.mk:
```diff
@@ -0,0 +1,35 @@
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
```
```suggestion
```

noroadsleft commented on keyboards/tpk/ace_of_spades/ace_of_spades.h:
```diff
@@ -0,0 +1,49 @@
+ * represents the switch matrix.
+ */
+
+#define LAYOUT_all( \
```
```suggestion
#define LAYOUT_tkl_iso_wkl( \
```

noroadsleft commented on keyboards/tpk/ace_of_spades/ace_of_spades.h:
```diff
@@ -0,0 +1,49 @@
+    { K500,  KC_NO, K502,  KC_NO, KC_NO, KC_NO, KC_NO, K505, KC_NO, KC_NO, K510,  KC_NO, K512,  K513,  KC_NO, K515,  K516  }, \
+}
+
+#define LAYOUT_tkl_iso LAYOUT_all
```
```suggestion
```

noroadsleft commented on keyboards/tpk/ace_of_spades/config.h:
```diff
@@ -0,0 +1,56 @@
+#define VENDOR_ID       0x504B
+#define PRODUCT_ID      0x1000
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    Matt
```
Any particular reason to not use `tpk` here?

noroadsleft commented on keyboards/tpk/ace_of_spades/keymaps/aholland909/keymap.c:
```diff
@@ -0,0 +1,57 @@
+
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [0] = LAYOUT_all(
```
```suggestion
  [0] = LAYOUT_tkl_iso_wkl(
```

noroadsleft commented on keyboards/tpk/ace_of_spades/keymaps/aholland909/keymap.c:
```diff
@@ -0,0 +1,57 @@
+    KC_LSFT,       KC_NUBS, KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT,          KC_UP,
+    KC_LCTL,                KC_LALT,                    KC_SPC,                                KC_RGUI,          TG(1),   KC_LEFT, KC_DOWN, KC_RGHT),
+
+  [1] = LAYOUT_all(
```
```suggestion
  [1] = LAYOUT_tkl_iso_wkl(
```

noroadsleft commented on keyboards/tpk/ace_of_spades/keymaps/aholland909/keymap.c:
```diff
@@ -0,0 +1,57 @@
+    _______,  KC_MPRV, KC_VOLD,  KC_MNXT,  _______,  _______,  KC_LEFT,  KC_DOWN,  KC_UP,    KC_RGHT,  _______,  _______,  _______, _______,
+    _______,  _______, _______,  _______,  KC_CALC,  _______,  _______,  _______,  _______,  _______,  _______,  _______,           _______,            RGB_HUI,
+    _______,           _______,                      KC_MPLY,                                _______,            _______,                     RGB_SAD,  RGB_HUD,  RGB_VAI),
+
+    [2] = LAYOUT_all(
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,          _______,
+    _______,          _______,                   _______,                            _______,          _______,                   _______, _______, _______),
+
+  [3] = LAYOUT_all(
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,          _______,
+    _______,          _______,                   _______,                            _______,          _______,                   _______, _______, _______),
```
These layers aren't needed here.

```suggestion
```

noroadsleft commented on keyboards/tpk/ace_of_spades/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,38 @@
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [0] = LAYOUT_all(
```
```suggestion
  [0] = LAYOUT_tkl_iso_wkl(
```

noroadsleft commented on keyboards/tpk/ace_of_spades/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,38 @@
+    KC_LSFT, KC_NUBS, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT,          KC_UP,
+    KC_LCTL,          KC_LALT,                   KC_SPC,                             KC_RGUI,          MO(1),                     KC_LEFT, KC_DOWN, KC_RGHT
+  ),
+  [1] = LAYOUT_all(
```
```suggestion
  [1] = LAYOUT_tkl_iso_wkl(
```

noroadsleft commented on keyboards/tpk/ace_of_spades/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+The Ace of Spades is a custom limited run TKL designed by tpk and manufactured by Salvun.
+
+Keyboard Maintainers: [aholland909](https://github.com/aholland909), [rys](https://github.com/rys)
+Hardware Supported: Ace of Spades custom PCB by [aholland909](https://github.com/aholland909)
+Hardware Availability: Custom group buy
```
```suggestion
* Keyboard Maintainers: [aholland909](https://github.com/aholland909), [rys](https://github.com/rys)
* Hardware Supported: Ace of Spades custom PCB by [aholland909](https://github.com/aholland909)
* Hardware Availability: Custom group buy
```

noroadsleft commented on keyboards/tpk/ace_of_spades/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+Make example for this keyboard (after setting up your build environment):
+
+`make tpk/ace_of_spades:default`
```
We actually prefer these as code blocks, instead of inline.

```suggestion
    make tpk/ace_of_spades:default
```

noroadsleft commented on keyboards/tpk/ace_of_spades/rules.mk:
```diff
@@ -0,0 +1,35 @@
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
+
+LAYOUTS = tkl_iso
```
This actually has a specific meaning in QMK, which doesn't apply to this keyboard, because the layout's different.

```suggestion
```

rys commented:
I've fixed all of the requested changes, and also renamed the board. The designer has decided to go with Ace of Spades as the whole brand now since he now plans others, and so this board is now called TKL.

So I moved `tpk` to `aos` in the tree, and `ace_of_spades` was moved to `tkl`. Updated the documentation to suit but let me know if I missed anything.

fauxpark commented:
Missing an info.json, which will be needed for Configurator support, but not strictly necessary for merge.

rys commented:
> Missing an info.json, which will be needed for Configurator support, but not strictly necessary for merge.

I've added one of those too in ae98d02. I'll add it to my checklist of things to provide for new boards.

noroadsleft commented:
Thanks!
